infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogCompaction.java

```diff
@@ -0,0 +1,52 @@
+package org.corfudb.infrastructure.log;
+
+import com.codahale.metrics.Counter;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.runtime.CorfuRuntime;
+
+import java.util.concurrent.*;
+
+/**
+ * Scheduled log compaction manager.
+ * Execute log compaction safely (prevent scheduled executor corruption if the task throws an exception)
+ * https://stackoverflow.com/questions/6894595/scheduledexecutorservice-exception-handling
+ */
+@Slf4j
+public class StreamLogCompaction {
+    public static final String STREAM_COMPACT_METRIC = StreamLogCompaction.class.getName();
+
+    private final ThreadFactory threadFactory = new ThreadFactoryBuilder()
+            .setDaemon(true)
+            .setNameFormat("LogUnit-Maintenance-%d")
+            .build();
+
+    /**
+     * How many times log compaction executed
+     */
+    private final Counter gc = CorfuRuntime.getDefaultMetrics().counter(STREAM_COMPACT_METRIC);
+
+    /**
+     * A scheduler, which is used to schedule periodic garbage collection.
+     */
+    private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor(threadFactory);
+    private final ScheduledFuture<?> compactor;
+
+    public StreamLogCompaction(StreamLog streamLog, long initialDelay, long period, TimeUnit timeUnit) {
+        Runnable task = () -> {
+            gc.inc();
+            log.debug("Start log compaction. GC counter: {}", gc.getCount());
+            try {
+                streamLog.compact();
+            } catch (Exception ex) {
+                log.error("Can't compact stream log. GC counter: {}", gc.getCount(), ex);
+            }
+        };
+        compactor = scheduler.scheduleWithFixedDelay(task, initialDelay, period, timeUnit);
+    }
+
+    public void shutdown() {
+        compactor.cancel(true);
+        scheduler.shutdownNow();
```

Comment:
awaitTermination of the executor service after shutdownNow. For example and timeouts to use see ManagementAgent.